### ISSUE TITLE
String with emoji in it throws off semantic highlighting

### ISSUE THREAD

**Issue by vadviktor** [Shopify/ruby-lsp#1162]
### Ruby version

3.1.4

### Code snippet

puts "🌱 string with emoji #{Rails.env}" }

### Description

I'm unsure if this is a known issue and/or if this is the LSP's issue. I still thought to report to you because you are the experts here 😅 .

I've tested this error by disabling ligatures and using Nerd Fonts; the results are unaffected.

I'm also using some custom TextMate rules, so I've disabled those, but the results were the same.

When there are no emojis in a string, the highlighting and other information features the LSP provides work just fine.

<img width="720" alt="Code_RIsByfeLKG" src="https://github.com/Shopify/ruby-lsp/assets/238703/2b902d66-7674-4985-b4c6-5c88c876c764">

When I add an emoji, the text analysis is broken:

<img width="665" alt="Code_94sXNq2hox" src="https://github.com/Shopify/ruby-lsp/assets/238703/257f5ce7-dd29-40c2-abdc-792de966f3db">

<img width="727" alt="Code_277qFtMNCX" src="https://github.com/Shopify/ruby-lsp/assets/238703/66c53dd1-756e-4d25-9955-7b3f7e97fe7b">

No matter what kind of emoji I'm using:

<img width="726" alt="Code_b17C2FNsQp" src="https://github.com/Shopify/ruby-lsp/assets/238703/f7dd6105-b1f8-48a8-ad04-5ae15ca30a33">

Also if I'm adding more emojis, it only seems to nudge the parsing to the right direction, only until I delete them or add more 😓 

<img width="687" alt="Code_rn2p2F8Qyo" src="https://github.com/Shopify/ruby-lsp/assets/238703/06de9ba5-bbe4-4f98-9062-e024d9d81866">

<img width="770" alt="Code_a0WYtBlmg0" src="https://github.com/Shopify/ruby-lsp/assets/238703/47aab259-9787-4003-8891-f0d88e98ca6e">

### Expected output

No matter how many emojis I use in a string, the semantic highlighting and other features provided by the LSP works the same way as with text and code without emoji.

**Comment by vinistock** [comment]
Thank you for the bug report! Yes, this is indeed an issue we need to solve for clients that are written in NodeJS.

The reason this happens is because the encoding is UTF-16, but NodeJS has an optimization for the string `length` method to return results in linear time, which means the positions are actually codepoints instead of characters (there's more context in #1006).

The fix here is that we always need to count UTF-16 length characters if the encoding is selected as UTF-16, since for NodeJS an emoji has length 2.

**Comment by sqwxl** [comment]
After some head scratching trying to understand why ruby-lsp was misbehaving, I tracked it down to this issue. I'm using some emoji's in my code. Turns out this does more than just throw off semantic highlighting, it basically breaks ruby-lsp. I've also been seeing 100% CPU runaway ruby-lsp process, but I'm not sure if that's related.

I understand the linked issue #1006 has been marked as not planned, but is there any workaround I can use here? Seems silly to be forced to drop emojis in 2025.
